flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/FineGrainedSlotManagerTestBase.java

```diff
@@ -0,0 +1,178 @@
+     */
+    protected abstract ResourceProfile getLargeSlotResourceProfile();
+
+    protected abstract Optional<ResourceAllocationStrategy> getResourceAllocationStrategy();
+
+    static SlotStatus createAllocatedSlotStatus(
+            AllocationID allocationID, ResourceProfile resourceProfile) {
+        return new SlotStatus(
+                new SlotID(ResourceID.generate(), 0), resourceProfile, new JobID(), allocationID);
+    }
+
+    static int getTotalResourceCount(Collection<ResourceRequirement> resources) {
+        if (resources == null) {
+            return 0;
+        }
+        return resources.stream()
+                .map(ResourceRequirement::getNumberOfRequiredSlots)
+                .reduce(0, Integer::sum);
+    }
+
+    static ResourceRequirements createResourceRequirementsForSingleSlot() {
+        return createResourceRequirementsForSingleSlot(new JobID());
+    }
+
+    static ResourceRequirements createResourceRequirementsForSingleSlot(JobID jobId) {
+        return createResourceRequirements(jobId, 1);
+    }
+
+    static ResourceRequirements createResourceRequirements(JobID jobId, int numRequiredSlots) {
+        return createResourceRequirements(jobId, numRequiredSlots, ResourceProfile.UNKNOWN);
+    }
+
+    static ResourceRequirements createResourceRequirements(
+            JobID jobId, int numRequiredSlots, ResourceProfile resourceProfile) {
+        return ResourceRequirements.create(
+                jobId,
+                "foobar",
+                Collections.singleton(
+                        ResourceRequirement.create(resourceProfile, numRequiredSlots)));
+    }
+
+    static TaskExecutorConnection createTaskExecutorConnection() {
+        return new TaskExecutorConnection(
+                ResourceID.generate(),
+                new TestingTaskExecutorGatewayBuilder().createTestingTaskExecutorGateway());
+    }
+
+    /** This class provides a self-contained context for each test case. */
+    protected class Context {
+        private final ResourceManagerId resourceManagerId = ResourceManagerId.generate();
+        private final ResourceTracker resourceTracker = new DefaultResourceTracker();
+        private final TaskManagerTracker taskManagerTracker = new FineGrainedTaskManagerTracker();
+        private final SlotStatusSyncer slotStatusSyncer =
+                new DefaultSlotStatusSyncer(Time.seconds(10L));
+        private final ScheduledExecutor scheduledExecutor = TestingUtils.defaultScheduledExecutor();
+        private final SlotManagerMetricGroup slotManagerMetricGroup =
+                UnregisteredMetricGroups.createUnregisteredSlotManagerMetricGroup();
+        private FineGrainedSlotManager slotManager;
+
+        protected TestingResourceAllocationStrategy.Builder resourceAllocationStrategyBuilder =
+                TestingResourceAllocationStrategy.newBuilder();
+        protected Executor mainThreadExecutor = MAIN_THREAD_EXECUTOR;
```

Comment:
Currently, some of the events are internally scheduled on the main thread executor, while others are directly called on the slot manager in the testing thread. We should make sure the slot manager is always accessed from the same thread.